lib/src/repo.rs

```diff
@@ -1220,6 +1234,42 @@ where
         IdIndex(vec)
     }
 
+    /// Looks up entries with the given prefix, and collects values if matched
+    /// entries have unambiguous keys.
+    pub fn resolve_prefix_with<U>(
+        &self,
+        prefix: &HexPrefix,
+        mut value_mapper: impl FnMut(&V) -> U,
+    ) -> PrefixResolution<Vec<U>> {
+        let mut range = self.resolve_prefix_range(prefix).peekable();
+        if let Some((first_key, _)) = range.peek().copied() {
+            let maybe_entries: Option<Vec<_>> = range
+                .map(|(k, v)| (k == first_key).then(|| value_mapper(v)))
+                .collect();
+            if let Some(entries) = maybe_entries {
+                PrefixResolution::SingleMatch(entries)
+            } else {
+                PrefixResolution::AmbiguousMatch
+            }
+        } else {
+            PrefixResolution::NoMatch
+        }
+    }
+
+    /// Iterates over entries with the given prefix.
+    pub fn resolve_prefix_range<'a: 'b, 'b>(
+        &'a self,
+        prefix: &'b HexPrefix,
+    ) -> impl Iterator<Item = (&'a K, &'a V)> + 'b {
+        let (bytes_prefix, min_bytes_prefix) = prefix.bytes_prefixes::<K>();
+        let pos = self.0.partition_point(|(k, _)| k < &min_bytes_prefix);
+        self.0[pos..]
+            .iter()
+            .take_while(move |(k, _)| k.as_bytes().starts_with(bytes_prefix.as_bytes()))
+            .filter(|(k, _)| prefix.matches(k))
```

Comment:
(See above) Why not just `take_while(|k| prefix.matches(k))`?

Reply:
For some reason, the comment this is related to (https://github.com/martinvonz/jj/pull/1124#discussion_r1087424756) ended up below this one.

Reply:
Good catch. I copied the logic from index.rs, but `bytes_prefix` test wouldn't be needed.
Maybe it's there because `prefix.matches()` was slower before?

I'll make another PR for this.